third_party/nvfuser/csrc/expr_simplifier.cpp

```diff
@@ -1633,87 +1647,76 @@ Val* reducePredicateRegisterUsage(Val* value, const Context& context) {
   if (!hasSimilarType(ltype, rtype)) {
     return value;
   }
-  // Redistribute terms. We always put general purposed registers to the
-  // left, and immediates to the right. If there is no immediates, then
-  // we put uniform registers to the right, otherwise we put uniform
-  // registers to the left.
-  bool moved = false;
-  std::vector<Val*> new_lhs;
-  std::vector<Val*> new_rhs;
-  std::vector<Val*> lhs_uniform;
-  std::vector<Val*> rhs_uniform;
-  auto neg = [&](Val* x) {
-    moved = true;
-    return IrBuilder::negExpr(x);
-  };
-  // Redistribute lhs_terms into new_lhs, new_rhs, and lhs_uniform, can only
-  // be called once.
-  auto redist_lhs = [&](const std::vector<Val*>& lhs_terms) {
-    for (auto inp : lhs_terms) {
-      if (prove::isZero(inp)) {
-        continue;
-      }
-      switch (getRegisterType(inp, context)) {
-        case RegisterType::GeneralPurpose:
-          new_lhs.emplace_back(inp);
-          break;
-        case RegisterType::Uniform:
-          lhs_uniform.emplace_back(inp);
-          break;
-        case RegisterType::Immediate:
-          new_rhs.emplace_back(neg(inp));
-          break;
-        default:
-          TORCH_INTERNAL_ASSERT(false, "Unknown register type");
-      }
-    }
-  };
-  // Redistribute rhs_terms into new_lhs, new_rhs, and rhs_uniform, can only
-  // be called once.
-  auto redist_rhs = [&](const std::vector<Val*>& rhs_terms) {
-    for (auto inp : rhs_terms) {
-      if (prove::isZero(inp)) {
+
+  // Classify terms into unroll and other
+  auto classify = [&context](const std::vector<Val*>& terms) {
+    std::vector<Val*> unroll, other;
+    RegisterType unroll_rtype = RegisterType::Unknown,
+                 other_rtype = RegisterType::Unknown;
+    for (auto inp : terms) {
+      if (inp->isZero()) {
         continue;
       }
-      switch (getRegisterType(inp, context)) {
-        case RegisterType::GeneralPurpose:
-          new_lhs.emplace_back(neg(inp));
-          break;
-        case RegisterType::Uniform:
-          rhs_uniform.emplace_back(inp);
-          break;
-        case RegisterType::Immediate:
-          new_rhs.emplace_back(inp);
-          break;
-        default:
-          TORCH_INTERNAL_ASSERT(false, "Unknown register type");
+      auto rtype = getRegisterType(inp, context);
+      if (hasUnrolledLoopIndex(inp, context)) {
+        unroll.push_back(inp);
+        unroll_rtype = promoteRegisterType(unroll_rtype, rtype);
+      } else {
+        other.push_back(inp);
+        other_rtype = promoteRegisterType(other_rtype, rtype);
       }
     }
+    return std::make_tuple(unroll, unroll_rtype, other, other_rtype);
   };
-  if (auto lhs_fop = toFlattenedAdd(bop->lhs()->definition())) {
-    redist_lhs(lhs_fop->inputs());
-  } else {
-    redist_lhs({bop->lhs()});
+
+  auto [lhs_unroll, lhs_unroll_rtype, lhs_other, lhs_other_rtype] = classify(
```

Comment:
I didn't know about this syntax. I'm assuming it has the same effect as using `std::tie`, right?

Reply:
It is similar, but not the same. This is called [structured binding](https://en.cppreference.com/w/cpp/language/structured_binding). My understanding is, structured binding is mostly used to declare and initialize a variable. `std::tie` is mostly for assignment.

Reply:
Oh, so this can be used for variable declarations. That sounds handy.